ncm-systemd/src/main/perl/Systemd/UnitFile.pm

```diff
@@ -0,0 +1,444 @@
+    $self->{log} = $opts{log} if $opts{log};
+
+    return SUCCESS;
+}
+
+=item custom
+
+The C<custom> method prepares configuration data that is cannot be
+found in the profile.
+
+Report hashref with custom data on success, undef otherwise.
+
+Following custom attributes are supported:
+
+=over
+
+=item CPUAffinity
+
+Obtain the C<systemd.exec> C<CPUAffinity> list determined via C<hwloc> locations.
+
+Allows to e.g. cpubind on numanodes using the C<node:X> location
+
+Forces an empty list to reset any possible previously defined affinity.
+
+=back
+
+=cut
+
+sub custom
+{
+    my ($self) = @_;
+
+    my $res = {};
+
+    foreach my $attr (keys %{$self->{custom}}) {
+        my $method = $CUSTOM_ATTRIBUTES{$attr};
+        if ($method) {
+            # the existence is unittested
+            if($self->can($method)) {
+                my $value = $self->$method($self->{custom}->{$attr});
+                if (defined($value)) {
+                    $res->{$attr} = $value;
+                } else {
+                    $self->error("Method $method for custom attribute $attr failed.");
+                    return;
+                }
+            } else {
+                $self->error("Unsupported method $method for custom attribute $attr.");
+                return;
+            }
+        } else {
+            $self->error("Unsupported custom attribute $attr.");
+            return;
+        }
+    }
+
+    return $res;
+}
+
+=item write
+
+Create the unitfile. Returns undef in case of problem,
+a boolean indication if something changed otherwise.
+
+(This method will take all required actions to use the values, like
+reloading the systemd daemon.
+It will not however change the state of the unit,
+e.g. by restarting it.)
+
+=cut
+
+sub write
+{
+    my ($self) = @_;
+
+    if (!(blessed($self->{config}) &&
+           $self->{config}->isa("EDG::WP4::CCM::Element"))) {
+        $self->error("config has to be an Element instance");
+        return;
+    }
+
+    # custom values
+    my $custom = $self->custom();
+    return if (! defined($custom));
+
+    # prepare/cleanup destination and return filename
+    my $filename = $self->_prepare_path($UNITFILE_DIRECTORY);
+    return if(! defined($filename));
+
+    # render
+    my $trd = EDG::WP4::CCM::TextRender->new(
+        $UNITFILE_TT,
+        $self->{config},
+        relpath => 'systemd',
+        log => $self,
+        ttoptions => _make_variables_custom($custom),
+        );
+
+    # write
+    my $fh = $trd->filewriter(
+        $filename,
+        backup => $self->{backup},
+        mode => 0664,
+        log => $self,
+        );
+
+    if(! defined($fh)) {
+        $self->error("Rendering unitfile for unit $self->{unit}",
+                     " (filename $filename) failed: $trd->{fail}.");
+        return;
+    }
+
+    my $changed = $fh->close() ? 1 : 0; # force to 1 or 0
+
+    # if changed, reload daemon
+    if($changed) {
+        # can't do much with return value?
```

Comment:
anything interesting in particular not logged in https://github.com/quattor/configuration-modules-core/pull/615/files#diff-dff1b898d6561f56515e6532273c7a28R166?